numpy/_array_api/_array_object.py

```diff
@@ -0,0 +1,891 @@
+            other = self._promote_scalar(other)
+        # Note: NumPy's __pow__ does not follow type promotion rules for 0-d
+        # arrays, so we use pow() here instead.
+        return pow(self, other)
+
+    def __rshift__(self: array, other: Union[int, array], /) -> array:
+        """
+        Performs the operation __rshift__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        # Note: The spec requires the return dtype of bitwise_right_shift, and
+        # hence also __rshift__, to be the same as the first argument.
+        # np.ndarray.__rshift__ returns a type that is the type promotion of
+        # the two input types.
+        res = self._array.__rshift__(other._array).astype(self.dtype)
+        return self.__class__._new(res)
+
+    def __setitem__(self, key, value, /):
+        """
+        Performs the operation __setitem__.
+        """
+        # Note: Only indices required by the spec are allowed. See the
+        # docstring of _validate_index
+        key = self._validate_index(key, self.shape)
+        res = self._array.__setitem__(key, asarray(value)._array)
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
+    def __sub__(self: array, other: Union[int, float, array], /) -> array:
+        """
+        Performs the operation __sub__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__sub__(other._array)
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
+    def __truediv__(self: array, other: Union[int, float, array], /) -> array:
+        """
+        Performs the operation __truediv__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__truediv__(other._array)
+        return self.__class__._new(res)
+
+    def __xor__(self: array, other: Union[int, bool, array], /) -> array:
+        """
+        Performs the operation __xor__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        self, other = self._normalize_two_args(self, other)
+        res = self._array.__xor__(other._array)
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
+    def __iadd__(self: array, other: Union[int, float, array], /) -> array:
+        """
+        Performs the operation __iadd__.
+        """
+        if isinstance(other, (int, float, bool)):
+            other = self._promote_scalar(other)
+        res = self._array.__iadd__(other._array)
+        if res.dtype != self.dtype:
+            raise RuntimeError
+        return self.__class__._new(res)
+
+    @np.errstate(all='ignore')
```

Comment:
This isn't directly part of the standard, but it's something we decided we should do, since the standard specifies specifically that various functions should return `nan` for various inputs. So basically the spirit of the standard is that there are no numerical errors. Error inputs give `nan`. 

Maybe @rgommers can comment on this too.

Reply:
Doesn't that effectively mean forcing non-IEEE conform floating point math without any consent?  Neither the library relying on the standard (and accepting a NumPy array), nor the end-user who passes in the NumPy array have a choice in warning about creation of new NaNs.

I can see that e.g. NaN creation warnings are often more pain than worth it, but I need a bit more convincing that this is well thought out here.

Reply:
After discussion with Ralf, I'm going to remove these `errstate` decorators. But to be sure, the spec does require returning nans in various places, and doesn't say anything about warnings. Is giving nans non-IEEE conforming? Maybe this issue should be brought up in the array-api repo.